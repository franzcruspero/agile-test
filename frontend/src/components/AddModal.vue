<template>
  <div>
    <div
      class="modal fade"
      :id="type"
      tabindex="-1"
      aria-labelledby="label"
      aria-hidden="true"
    >
      <div class="modal-dialog">
        <div class="modal-content">
          <div class="modal-header">
            <h5 class="modal-title" id="label">Add Value/Principle</h5>
            <button
              type="button"
              class="btn-close"
              data-bs-dismiss="modal"
              aria-label="Close"
            ></button>
          </div>
          <form
            @submit.prevent="submitForm"
            class="needs-validation"
            novalidate
          >
            <div class="modal-body">
              <div>
                <label for="title" class="form-label">Title</label>
                <input
                  type="text"
                  class="form-control"
                  id="title"
                  placeholder=""
                  v-model="title"
                  required
                />
                <div class="invalid-feedback">
                  Valid first name is required.
                </div>
              </div>
              <div>
                <label for="description" class="form-label">Description</label>
                <textarea
                  type="text"
                  class="form-control"
                  id="description"
                  placeholder=""
                  v-model="description"
                  rows="3"
                  required
                />
                <div class="invalid-feedback">
                  Valid first name is required.
                </div>
              </div>
            </div>
            <div class="modal-footer">
              <button
                type="button"
                class="btn btn-secondary"
                data-bs-dismiss="modal"
              >
                Close
              </button>
              <button
                type="submit"
                class="btn btn-primary"
                data-bs-dismiss="modal"
              >
                Save changes
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import axios from "axios";

export default {
  props: {
    type: String,
    list_reload: Number,
  },
  data() {
    return {};
  },
  methods: {
    async submitForm() {
      await axios.post(`http://localhost:8000/api/v1/${this.type}/`, {
        title: this.title,
        description: this.description,
      });

      this.title = "";
      this.description = "";
      this.$emit("submitForm");
    },
  },
};
</script>
