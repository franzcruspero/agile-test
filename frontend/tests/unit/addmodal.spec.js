import AddModal from "@/components/AddModal.vue"
import { mount } from "@vue/test-utils"
import axios from "axios"

jest.mock('axios', () => ({
    post: jest.fn(() => {})
}))


describe("addModal.vue", () => {
    let wrapper = mount(AddModal, {
        data() {
            return {
                type: "values",
                title: "initial value",
                description: "initial description"
            }
        }
    })

    it("shows modal total", () => {
        expect(wrapper.find("h5").text()).toBe("Add Value/Principle")
    })

    it("mocks axios post request and submit function", async() => {
        const title = "Test title"
        const description = "Test description"

        await wrapper.find("input").setValue(title)
        await wrapper.find("textarea").setValue(description)
        await wrapper.find('form').trigger('submit.prevent')

        expect(axios.post).toHaveBeenCalledTimes(1)

        const titleField = await wrapper.get("input")
        const descriptionField = await wrapper.get("textarea")

        expect(titleField.text()).toBe("")
        expect(descriptionField.text()).toBe("")

        expect(wrapper.emitted()).toHaveProperty("submitForm")


    })
    it('renders correctly', () => {
        const wrapper = mount(AddModal)
        expect(wrapper.element).toMatchSnapshot()
    })
})