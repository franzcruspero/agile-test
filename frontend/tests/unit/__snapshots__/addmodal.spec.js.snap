// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addModal.vue renders correctly 1`] = `
<div>
  <div
    aria-hidden="true"
    aria-labelledby="label"
    class="modal fade"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
            id="label"
          >
            Add Value/Principle
          </h5>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <form
          class="needs-validation"
          novalidate=""
        >
          <div
            class="modal-body"
          >
            <div>
              <label
                class="form-label"
                for="title"
              >
                Title
              </label>
              <input
                class="form-control"
                id="title"
                placeholder=""
                required=""
                type="text"
              />
              <div
                class="invalid-feedback"
              >
                 Valid first name is required. 
              </div>
            </div>
            <div>
              <label
                class="form-label"
                for="description"
              >
                Description
              </label>
              <textarea
                class="form-control"
                id="description"
                placeholder=""
                required=""
                rows="3"
                type="text"
              />
              <div
                class="invalid-feedback"
              >
                 Valid first name is required. 
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
               Close 
            </button>
            <button
              class="btn btn-primary"
              data-bs-dismiss="modal"
              id="submit-button"
              type="submit"
            >
               Save changes 
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
